pkg/vsphere/datastore/datastore.go

```diff
@@ -268,54 +252,56 @@ func (d *Helper) IsVSAN(ctx context.Context) bool {
 // result so the URI doesn't need to be recomputed for every datastore
 // operation.
 func (d *Helper) mkRootDir(ctx context.Context, rootdir string) error {
+	if rootdir == "" {
+		return fmt.Errorf("root directory is empty")
+	}
+
+	if path.IsAbs(rootdir) {
+		return fmt.Errorf("root directory (%s) must not be an absolute path", rootdir)
+	}
 
 	// Handle vsan
 	// Vsan will complain if the root dir exists.  Just call it directly and
 	// swallow the error if it's already there.
 	if d.IsVSAN(ctx) {
+		comps := strings.Split(rootdir, "/")
+
 		nm := object.NewDatastoreNamespaceManager(d.s.Vim25())
 
 		// This returns the vmfs path (including the datastore and directory
 		// UUIDs).  Use the directory UUID in future operations because it is
 		// the stable path which we can use regardless of vsan state.
-		uuid, err := nm.CreateDirectory(ctx, d.ds, rootdir, "")
+		uuid, err := nm.CreateDirectory(ctx, d.ds, comps[0], "")
 		if err != nil {
 			if !soap.IsSoapFault(err) {
 				return err
 			}
 
 			soapFault := soap.ToSoapFault(err)
-			_, ok := soapFault.VimFault().(types.FileAlreadyExists)
-			if ok {
+			if _, ok := soapFault.VimFault().(types.FileAlreadyExists); ok {
 				// XXX UGLY HACK until we move this into the installer.  Use the
 				// display name if the dir exists since we can't get the UUID after the
 				// directory is created.
 
-				uuid = rootdir
+				uuid = comps[0]
 				err = nil
 			} else {
 				return err
 			}
 		}
 
-		// set the root url to the UUID of the dir we created
-		d.RootURL = d.ds.Path(path.Base(uuid))
-		log.Infof("Created store parent directory (%s) at %s", rootdir, d.RootURL)
-	} else {
+		rootdir = path.Join(path.Base(uuid), path.Join(comps[1:]...))
+	}
 
-		// Handle regular local datastore
-		// check if it already exists
+	rooturl := d.ds.Path(rootdir)
 
-		d.RootURL = d.ds.Path(rootdir)
-		if _, err := d.Mkdir(ctx, true); err != nil {
-			if os.IsExist(err) {
-				log.Debugf("%s already exists", d.RootURL)
-				return nil
-			}
-			return err
-		}
+	// create the rest of the root dir in case of vSAN, otherwise
+	// create the full path
+	if _, err := mkdir(ctx, d.s, d.fm, true, rooturl); err != nil && !os.IsExist(err) {
+		return err
```

Comment:
George pointed out we're checking for the exist err.  Ignore my previous question.